pinot-integration-tests/src/test/java/org/apache/pinot/server/realtime/ControllerLeaderLocatorIntegrationTest.java

```diff
@@ -0,0 +1,169 @@
+    enableResourceConfigForLeadControllerResource(true);
+
+    // After resource config is enabled, use the lead controller in the resource.
+    validateResultSet(controllerLeaderLocator, resultSet, 1, "Failed to get only one pair of controller");
+
+    ControllerConf secondControllerConfig = getDefaultControllerConfiguration();
+    secondControllerConfig.setControllerPort(Integer.toString(DEFAULT_CONTROLLER_PORT + 1));
+    ControllerStarter secondControllerStarter = new ControllerStarter(secondControllerConfig);
+    secondControllerStarter.start();
+
+    TestUtils
+        .waitForCondition(aVoid -> secondControllerStarter.getHelixResourceManager().getHelixZkManager().isConnected(),
+            TIMEOUT_IN_MS, "Failed to start the second controller");
+
+    // After starting a second controller, there should be two leaders for all the partitions.
+    validateResultSet(controllerLeaderLocator, resultSet, 2, "Failed to get two pairs of controllers.");
+
+    // Disable lead controller resource
+    enableResourceConfigForLeadControllerResource(false);
+
+    // After disabling lead controller resource, only helix leader should be used.
+    validateResultSet(controllerLeaderLocator, resultSet, 1, "Failed to get only one pair of controller");
+
+    // Mock the behavior that 60 seconds have passed.
+    controllerLeaderLocator.setCurrentTimeMs(
+        controllerLeaderLocator.getCurrentTimeMs() + 2 * controllerLeaderLocator.getMinInvalidateIntervalMs());
+    controllerLeaderLocator.invalidateCachedControllerLeader();
+
+    // Randomly pick a table name to get a controller leader, and this leader should be the same as the helix leader.
+    Pair<String, Integer> pair = controllerLeaderLocator.getControllerLeader(
+        _partitionToTableMap.get(new Random().nextInt(Helix.NUMBER_OF_PARTITIONS_IN_LEAD_CONTROLLER_RESOURCE)));
+
+    Assert.assertEquals(pair.getFirst(), ControllerTest.LOCAL_HOST);
+    Assert.assertEquals((int) pair.getSecond(), ControllerTest.DEFAULT_CONTROLLER_PORT);
+
+    // Stop the second controller.
+    secondControllerStarter.stop();
+  }
+
+  /**
+   * Find the table names for all the partitions.
+   */
+  private void findTableNamesForAllPartitions() {
+    _partitionToTableMap = new HashMap<>();
+
+    int count = 0;
+    while (_partitionToTableMap.size() < Helix.NUMBER_OF_PARTITIONS_IN_LEAD_CONTROLLER_RESOURCE) {
+      String tableName = "testTable" + count;
+      int partitionId = LeadControllerUtils.getPartitionIdForTable(tableName);
+      _partitionToTableMap.putIfAbsent(partitionId, tableName);
+      count++;
+    }
+  }
+
+  private void validateResultSet(FakeControllerLeaderLocator controllerLeaderLocator, Set<String> resultSet,
+      int expectedNumberOfUniqueResults, String errorMessage) {
+    TestUtils.waitForCondition(aVoid -> {
+      resultSet.clear();
+      for (Map.Entry<Integer, String> entry : _partitionToTableMap.entrySet()) {
+        // Mock the behavior that 60 seconds have passed.
```

Comment:
nit: "Mock time so  it is beyond minimum time to invalidate leader cache"

Reply:
Done.